### PR TITLE
[RISC-V] Fix Int32 to Unsigned overflow check

### PR DESCRIPTION
**Pull request by Bajtazar**
* Fixes `int` to `uint` and `int` to `ulong` convertion overflow check on RISC-V. Previously the `blt a0, zero, RAISE_OVERFLOW` was generated, but the integer inside the source register was not sign extended resulting in this check always passing. The sign extending instruction was added before the branch in order to properly sign extend the incoming integer
* Added `ret` pseudoinstruction to RISC-V disasm
* Improved disasm of `j` pseudoinstruction
* Optimized other integer casts a little bit

Part of #84834, cc @dotnet/samsung

### CODE REVIEW

**Comment by risc-vv** [comment]
<details><summary>87105a9c is being scheduled for building and testing</summary>

GIT: `87105a9ccb8b1a0009350f376bfab36b1263058e`
REPO: [Bajtazar/runtime](https://github.com/Bajtazar/runtime)
BRANCH: riscv-positive-int-overflow-fix</details>

File: src/coreclr/jit/instr.cpp [hunk 1]
```diff
@@ -1861,9 +1861,12 @@ instruction CodeGenInterface::ins_Load(var_types srcType, bool aligned /*=false*
...
                 ins = INS_lh;
         }
-        else if (TYP_INT == srcType)
+        else if (varTypeIsInt(srcType))
         {
-            ins = INS_lw;
+            if (varTypeIsUnsigned(srcType))
+                ins = INS_lwu;
+            else
+                ins = INS_lw;
```
Comment (tomeksowi) [inline src/coreclr/jit/instr.cpp]: Small ints on RISC-V are type-extended to 32 bits, then sign-extended to full register length. So sign-extending uint was correct.

Comment (jakobbotsch) [inline src/coreclr/jit/instr.cpp:1869]: `TYP_UINT` is not a possible source type for loads in JIT IR. The previous code made more sense to me.

Comment (Bajtazar) [reply src/coreclr/jit/instr.cpp]: Ok, revoked this change. Thanks!

**Comment by risc-vv** [comment]
<details><summary>4bef7fbc is being scheduled for building and testing</summary>

GIT: `4bef7fbc355da58fc7851826ad6534db5e1f70bb`
REPO: [Bajtazar/runtime](https://github.com/Bajtazar/runtime)
BRANCH: riscv-positive-int-overflow-fix</details>

File: src/coreclr/jit/codegenriscv64.cpp [hunk 2]
```diff
@@ -6227,17 +6227,19 @@ void CodeGen::genJmpPlaceVarArgs()
...
 void CodeGen::genIntCastOverflowCheck(GenTreeCast* cast, const GenIntCastDesc& desc, regNumber reg)
 {
+    const regNumber tempReg = internalRegisters.GetSingle(cast);
+
     switch (desc.CheckKind())
     {
         case GenIntCastDesc::CHECK_POSITIVE:
         {
-            genJumpToThrowHlpBlk_la(SCK_OVERFLOW, INS_blt, reg, nullptr, REG_R0);
+            GetEmitter()->emitIns_R_R(INS_sext_w, EA_8BYTE, tempReg, reg);
```
Comment (tomeksowi) [inline src/coreclr/jit/codegenriscv64.cpp:6236]:
```suggestion
            GetEmitter()->emitIns_R_R(INS_sext_w, EA_4BYTE, tempReg, reg);
```
Nit. The parameter is unused, but *w instructions are EA_4BYTE everywhere.

**Comment by tomeksowi** [comment]
> * the integer inside the source register was not sign extended

Where did the unextended int come from?

File: src/coreclr/jit/codegenriscv64.cpp [hunk 3]
```diff
@@ -6267,7 +6268,6 @@ void CodeGen::genIntCastOverflowCheck(GenTreeCast* cast, const GenIntCastDesc& d
 
         case GenIntCastDesc::CHECK_INT_RANGE:
         {
-            const regNumber tempReg = internalRegisters.GetSingle(cast);
             assert(tempReg != reg);
             GetEmitter()->emitLoadImmediate(EA_8BYTE, tempReg, INT32_MAX);
             genJumpToThrowHlpBlk_la(SCK_OVERFLOW, INS_blt, tempReg, nullptr, reg);
```
Comment (tomeksowi) [inline src/coreclr/jit/codegenriscv64.cpp:6273]: Can `reg` be also badly extended here (or in other checks)?

**Comment by risc-vv** [comment]
<details><summary>bc09d23d is being scheduled for building and testing</summary>

GIT: `bc09d23d4b310b7878e9db4a38ebc5f80663e4b8`
REPO: [Bajtazar/runtime](https://github.com/Bajtazar/runtime)
BRANCH: riscv-positive-int-overflow-fix</details>

**Comment by risc-vv** [comment]
<details><summary>44e74115 is being scheduled for building and testing</summary>

GIT: `44e74115d413c1883de862f042aad21214e2245b`
REPO: [Bajtazar/runtime](https://github.com/Bajtazar/runtime)
BRANCH: riscv-positive-int-overflow-fix</details>

**Comment by risc-vv** [comment]
<details><summary>5fdc67f2 is being scheduled for building and testing</summary>

GIT: `5fdc67f20569fe496fb86a68820f949259f26f85`
REPO: [Bajtazar/runtime](https://github.com/Bajtazar/runtime)
BRANCH: riscv-positive-int-overflow-fix</details>

**Comment by risc-vv** [comment]
<details><summary>209a8b16 is being scheduled for building and testing</summary>

GIT: `209a8b16d938c9a7d723c4c45a8e7acb4d15ce0a`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Review by tomeksowi — APPROVED** [review]

File: src/coreclr/jit/emitriscv64.cpp [hunk 4]
```diff
@@ -3946,7 +3953,16 @@ void emitter::emitDispInsName(
...
-                printf("j              %d", offset);
+                printf("j              ");
+
+                if (id->idIsBound())
+                {
+                    emitPrintLabel(id->idAddr()->iiaIGlabel);
+                }
+                else
+                {
+                    printf("pc%+d instructions", offset >> 2);
```
Comment (clamp03) [inline src/coreclr/jit/emitriscv64.cpp:3964]: Please just set to offset like the others.

File: src/coreclr/jit/codegenriscv64.cpp [hunk 5]
```diff
@@ -6227,90 +6227,81 @@ void CodeGen::genJmpPlaceVarArgs()
...
+            {
+                // Check if bits leading the actual small int are all zeros
+                // If destination type is signed then also check if MSB of it is zero
+                const bool     isDstSigned = !varTypeIsUnsigned(cast->gtCastType);
+                const unsigned excludeMsb  = isDstSigned ? 1 : 0;
+                const unsigned typeSize    = 8 * castSize - excludeMsb;
+                GetEmitter()->emitIns_R_R_I(INS_srli, EA_8BYTE, tempReg, reg, typeSize);
+                genJumpToThrowHlpBlk_la(SCK_OVERFLOW, INS_bne, tempReg, nullptr, REG_R0);
+            }
+            else // Signed to Signed convertion
```
Comment (clamp03) [inline src/coreclr/jit/codegenriscv64.cpp:6298]: typo

Comment (Bajtazar) [reply src/coreclr/jit/codegenriscv64.cpp]: Done

**Comment by risc-vv** [comment]
<details><summary>97c951c6 is being scheduled for building and testing</summary>

GIT: `97c951c6832b702b0ebf7808c06e0073beb1c0e6`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Review by clamp03 — APPROVED** [review]

**Review by yurai007 — APPROVED** [review]

**Comment by risc-vv** [comment]
<details><summary>daa73aa3 is being scheduled for building and testing</summary>

GIT: `daa73aa37617aa118b1a4dde3d3bc3a20619476a`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Review by sirntar — APPROVED** [review]

**Comment by Bajtazar** [comment]
@jakobbotsch @jkotas Could you please merge this? I don't think there is anything more left to address in this PR

**Comment by jakobbotsch** [comment]
Sorry, I was on vacation (if you can, tagging @dotnet/jit-contrib would get attention from more people).
Going to bounce the CI here...

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**